web/testing/issue917-context-and-notany/index.html

```diff
@@ -0,0 +1,107 @@
+      
+    <title>KeymanWeb Testing Page - context() and notany() interaction (#917)</title>
+
+    <!-- Your page CSS --> 
+    <style type='text/css'>   
+      body {font-family: Tahoma,helvetica;}
+      h3 {font-size: 1em;font-weight:normal;color: darkred; margin-bottom: 4px}
+      .test {font-size: 1.5em; width:80%; min-height:30px; border: 1px solid gray;}
+      #KeymanWebControl {width:50%;min-width:600px;}       
+    </style> 
+ 
+    <!-- Insert uncompiled KeymanWeb source scripts -->              
+    <script src="../../release/unminified/web/keymanweb.js" type="application/javascript"></script>
+
+    <!-- 
+      For desktop browsers, a script for the user interface must be inserted here.
+       
+      Standard UIs are toggle, button, float and toolbar.  
+      The toolbar UI is best for any page designed to support keyboards for 
+      a large number of languages.
+    -->
+    <script src="../../release/unminified/web/kmwuitoggle.js"></script>
+ 
+    <!-- Initialization: set paths to keyboards, resources and fonts as required -->
+    <script>
+      var kmw=window.keyman;
+      kmw.init({
+		    attachType:'auto'
+      }).then(function() {
+        kmw.addKeyboards({id:'test_917',name:'test_917',languages:{id:'en',name:'English'}, filename:'test_917/build/test_917.js'});
+      });
+
+    </script>  
+  </head>
+
+<!-- Sample page HTML -->  
+  <body> 
+    <h2>KeymanWeb Sample Page -context() and notany() interaction (<a href="https://github.com/keymanapp/keyman/issues/917">#917</a>)</h2>
+    <p>KeymanWeb did not support notany() and context() together. This keyboard tests various scenarios in the implementation (14.0 min):</p>
+    <table>
+      <thead><tr><th>Input</th><th>Result</th><th>Notes</th></tr></thead>
+      <tbody>
+        <tr><td>abc</td><td>'abc'</td><td>Should not match rule on line 19</td></tr>
+        <tr><td>xbc</td><td>'xBC'</td><td>Basic character match at start of context (notany vowel) on line 19</td></tr>
+        <tr><td>dbc</td><td>DK(1) 'BC'</td><td>Deadkey match at start of context on line 19</td></tr>
+
+        <tr><td>adcd</td><td>'a' DK(1) 'XX'</td><td>Second position in context, match deadkey rule on line 27</td></tr>
+        <tr><td>adcd?</td><td>'pass!'</td><td>Verifies that deadkey is conserved through rule match on line 27</td></tr>
+        <tr><td>axcd</td><td>'axXX'</td><td>Verifies that match occurs for character on line 27</td></tr>
+        <tr><td>axcd?</td><td>'axXX?'</td><td>Sanity check for line 27</td></tr>
+
+        <tr><td>fghi</td><td>'fgXX'</td><td>Testing adjusted index for `if` statement on line 35</td></tr>
+        <tr><td>fghi?</td><td>'fgXX?'</td><td>Sanity check for line 35</td></tr>
+        <tr><td>fdhi</td><td>'f' DK(1) XX'</td><td>Testing deadkey match for adjusted index for `if` statement on line 35</td></tr>
+        <tr><td>fdhi?</td><td>'pass!'</td><td>Verifies that deadkey is conserved thrghou rule match on line 35</td></tr>
```

Comment:
Manual tests that could be converted into automatic tests...

Reply:
The most applicable site for this would be found at the following path:  [`common/core/web/keyboard-processor/tests/cases/engine/context.js`](https://github.com/keymanapp/keyman/blob/master/common/core/web/keyboard-processor/tests/cases/engine/context.js).

The test keyboard should be placed within `common/core/web/tests/resources/keyboards`, as that's where a number of common web test resources are stored.  (In particular, the resources already referenced by the tests within the site mentioned above.)

The patterns for existing unit tests within that file are quite well-structured, so I don't expect much difficulty with a conversion.  Feel free to either include them within that file or write a separate but similarly structured file.  Alternatively, I should be able to do the conversion once I'm in a good place to spare a couple of hours.

Reply:
Probably something separate, as the actual rule application (`runEngineRuleSet`) defined in that file is very tailored to `fullContextMatch`.  Aside from that, and possibly some details of the test spec pattern, the rest of the file should be pretty applicable.

For programatically emulating keystrokes (as seen within `runEngineRuleSet`):

```
      var matchTest = matchDefs[j];
      var ruleSeq = new KMWRecorder.RecordedKeystrokeSequence(matchTest.sequence);
      let proctor = new KMWRecorder.NodeProctor(keyboard, device, assert.equal);

      // We want to specify the OutputTarget for this test; our actual concern is the resulting context.
      var target = new com.keyman.text.Mock();
      ruleSeq.test(proctor, target);
```

The final line in that block executes the keystroke sequence defined by each spec's `.sequence` property.  In fact, if all you want to do is check against equality with the intended output, that's the core part you'd need; everything afterward was established for extra-thorough checks against `fullContextMatch`, which won't matter for you.  Just the `sequence` definitions should suffice for automating these tests.

Reply:
Tests created in notany_context.js 💯